exp/etcdrestore/controllers/etcdsnaphotsync_controller.go

```diff
@@ -51,6 +64,107 @@ func (r *EtcdSnapshotSyncReconciler) SetupWithManager(_ context.Context, mgr ctr
 	return nil
 }
 
-func (r *EtcdSnapshotSyncReconciler) Reconcile(_ context.Context, _ ctrl.Request) (res ctrl.Result, reterr error) {
+//+kubebuilder:rbac:groups=turtles-capi.cattle.io,resources=rke2etcdmachinesnapshotconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=turtles-capi.cattle.io,resources=rke2etcdmachinesnapshotconfigs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=turtles-capi.cattle.io,resources=rke2etcdmachinesnapshotconfigs/finalizers,verbs=update
+
+func (r *EtcdSnapshotSyncReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := log.FromContext(ctx)
+
+	log.Info("Reconciling CAPI cluster and syncing etcd snapshots")
+
+	cluster := &clusterv1.Cluster{}
+	if err := r.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if cluster.Spec.Paused {
+		log.Info("Cluster is paused, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	// Only reconcile RKE2 clusters
+	if cluster.Spec.ControlPlaneRef.Kind != RKE2ControlPlaneKind { // TODO: Move to predicate
+		log.Info("Cluster is not an RKE2 cluster, skipping reconciliation")
+		return ctrl.Result{RequeueAfter: 3 * time.Minute}, nil
+	}
+
+	if !cluster.Status.ControlPlaneReady {
+		log.Info("Control plane is not ready, skipping reconciliation")
+		return ctrl.Result{RequeueAfter: 3 * time.Minute}, nil
+	}
+
+	if err := r.watchEtcdSnapshotFiles(ctx, cluster); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to start watch on ETCDSnapshotFile: %w", err)
+	}
+
+	return r.reconcileNormal(ctx, cluster)
+}
+
+func (r *EtcdSnapshotSyncReconciler) reconcileNormal(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+	remoteClient, err := r.Tracker.GetClient(ctx, capiutil.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var snapshotter snapshotters.Snapshotter
+
+	switch cluster.Spec.ControlPlaneRef.Kind {
+	case RKE2ControlPlaneKind:
+		snapshotter = snapshotters.NewRKE2Snapshotter(r.Client, remoteClient, cluster)
+	default:
+		return ctrl.Result{}, fmt.Errorf("unsupported control plane kind: %s", cluster.Spec.ControlPlaneRef.Kind)
+	}
+
+	if err := snapshotter.Sync(ctx); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to sync etcd snapshots: %w", err)
+	}
+
 	return ctrl.Result{}, nil
 }
+
+func (r *EtcdSnapshotSyncReconciler) watchEtcdSnapshotFiles(ctx context.Context, cluster *clusterv1.Cluster) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(5).Info("Setting up watch on ETCDSnapshotFile")
+
+	etcdnapshotFile := &unstructured.Unstructured{}
+	etcdnapshotFile.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "k3s.cattle.io",
+		Kind:    "ETCDSnapshotFile",
+		Version: "v1",
+	})
```

Comment:
Does remote tracker support watching typed ETCDSnapshotFile? Maybe we need to bump CAPI to latest version to allow this?

Reply:
yes, I forgot to update this part after we made a wrapper for ETCDSnapshotFile